man/man3/acl_from_text.3
.TH ACL_FROM_TEXT 3
.SH NAME
acl_from_text, acl_to_short_text, acl_to_text \- convert a POSIX ACL string to a struct acl or
a struct acl to a POSIX ACL string 
.SH SYNOPSIS
.B #include <sys/acl.h>
.PP
.B acl_t acl_from_text(const char *argp);
.br
.B char * acl_to_short_text(struct acl * aclp, ssize_t *len_p);
.br
.B char * acl_to_text(struct acl * aclp, ssize_t *len_p);
.SH DESCRIPTION
These routines convert strings defined by the POSIX P1003.1e specifications
(see
.IR xfs_acl (1))
to/from 
.IR "struct acl" , 
which is the internal
format for an
Access Control List.  
.SH RETURN VALUES
\f2acl_from_text\fP returns a pointer to a \f2struct acl\fP
allocated by the routine or a NULL pointer in the case of a failure.
The acl should be freed when no longer needed by calling 
.IR acl_free (3).
.PP
\f2acl_to_text\fP returns a pointer to a null terminated 
character string allocated by the routine which is
a long form representation of the ACL, or a NULL pointer if an error
occurred.  This string
should be freed when no longer needed by calling
.IR acl_free (3).
If the \f2len_p\fP argument is not NULL, then the length of the string
is returned in \f2ssize_t\fP pointed to by \f2len_p\fP.
.PP
\f2acl_to_short_text\fP is identical to \f2acl_to_text\fP, except that a
short form representation of the ACL is returned.
.SH ERRORS
.I acl_from_text:
.TP 16
EINVAL
input text is invalid
.TP 16
ENOMEM
could not allocate space for new acl
.PP
.I acl_to_short_text:
.TP 16
EINVAL 
input acl is invalid
.TP 16
ENOMEM 
could not allocate space for the return string
.PP
.I acl_to_text:
.TP 16
EINVAL 
input acl is invalid
.TP 16
ENOMEM 
could not allocate space for the return string


doc/CHANGES
2.0.6 (08 April 2002)
* Fix a bug that causes setfacl not to report error messages if
  acl-set_file() fails (the bug was introduced in version 2.0.3).

2.0.5 (26 March 2002)
* Fix up acl_to_text compliance with POSIX draft.
* Add the new acl(5) manual page.
* Add a manual section 3 directory and Makefile.
* Add the libacl manual pages.
* Rename the existing acl_to_any_text function to __acl_to_any_text, and
  add a new acl_to_any_text function that only contains parameters that
  are actually useful.  Make acl_to_text and acl_to_any_text invoke
  __acl_to_any_text internally.  Update all uses of acl_to_any_text in
  getfacl/getfacl.c, setfacl/do_set.c, and chacl/chacl.c.
* Move the acl_size function form the LIBACL_CFILES to the POSIX_CFILES
  section (it is a POSIX function)
* Rearrange libacl routines so that several non-POSIX routines which
  were being exported are no longer exported (eg. acl_entry_to_any_str,
  acl_get_file_mode, acl_get_fd_mode, acl_set_file_mode, acl_set_fd_mode).

2.0.4 (12 March 2002)
* Man page updates from Andreas
* Test script updates from Andreas

2.0.3 (05 March 2002)
* Clean up the --default option handling in setfacl. The old
  workarounds caused a bug for unusual input.
* Changes to the --test output format setfacl generates: ACLs that
  are not changed are now displayed as `*'.
* Fix a bug in setfacl/sequence.c:seq_delete_cmd().
* Minor changes to test scripts

2.0.2 (01 March 2002)
* Apply several patches from Andreas, namely:
* man page fixes
* libacl code reformatting
* acl_from_text errno handling

2.0.1 (27 February 2002)
* Apply Andreas Gruenbacher's diffs.
* Fix up chacl for deletion of access ACL to be in line with Andreas. 
* Incorporate the Debian packaging again.

2.0.0 (21 February 2002)
* Reworked to use the new official system call API.
* Sync up with the XFS project, the SGI folk now use this source.
* Jumped to version 2 to allow XFS users to upgrade
  (Rationale: the XFS ACL user tools were at version 1.1.X, and
  packaging tools like rpm, dpkg, etc. must be presented with a
  greater version number to allow an upgrade to proceed).
* Added the chacl command to ease migration for existing XFS users,
  and for compatibility with IRIX.
* Added a flag to allow acl_print to produce a single-line ACL, in
  addition to the multi-line format.
* Extended attribute documentation has moved into the extended
  attribute package from SGI ("attr"), this ACL package now deals
  exclusively with ACLs.
* acl_from_text sometimes did not set errno when failing.
* Moved files and simplified #includes in libacl

0.7.22
* Create a shared version of libacl, instead of linking statically.
* Added missing error reporting to getfacl do_print() function.     

0.7.16
* Minor cleanups/fixes in the make files
* The test cases in test/src have been updated to match the
  corrected ACL entry syntax (see the 0.7.15 changes).

0.7.15
* acl_from_text() now accepts mask and other entries with either two
  or three fields (`other:rwx' or `other::rwx'). Posix 1003.1e specifies
  that these entries should have three entries, while under Solaris,
  they have only two fields.
* setfacl now parses mask and other entries with two or three entries.
* getfacl now produces mask and other entries with three fields
  (`mask::rwx' instead of the previous `mask:rwx').
* acl_get_fd(), acl_get_file(), acl_set_fd(), acl_set_file() now only
  fail with errno=ENOSYS if the underlying filesystem doesn't support
  ACLs. acl_get_fd() and acl_get_file() construct ACLs from the file
  mode permission bits if the kernel supports ACLs, but no ACL extended
  attributes are defined for the file (errno=ENOATTR).
* acl_to_text() wrongly separated ACL entires with ',' instead of '\n'.
* acl_from_text() only accepted three-letter permission strings
  (rwx, ---, rw-, etc.). Now also short strings (rw, -, etc.) are
  supported.

0.7.10
* acl_get_fd(), acl_get_file() now don't return basic ACL's anymore.
* acl_set_fd(), acl_set_file() fail on filesystems that don't support
  ACLs, even if the ACL to be set is a basice ACL.
* If msgfmt is not present on a system, the message catalog(s) are
  no longer being built.

0.7.8 (19 February 2001)
* Buffer overflow bug in ACL library (acl_to_any_text).
  (Reported by Charles Bertsch <CBertsch@microtest.com>)

0.7.2 (22 November 2000)
* Replace rman with groff for building HTML versions of the
  manual pages.
* Add descriptions of the permissions required for manipulating
  ACLs to the manual pages.
* Add long options to getfacl and setfacl, update the manual pages.
* Remove some small bugs from the src/showacl Perl script.
* showacl is now integrated in getfacl.
* Bug in acl_to_any_text() for ACLs with 0 entries fixed.
* Bug in acl_to_text(): Used '\n' as the ACL entry separator.
* German messages updated.

0.7.1 (23 October 2000)
* Basic GNU gettext support and a German translation of messages
  added.

0.7.0 (22 October 2000)
* Updates to the documentation
* Now an ACL_MASK entry is cloned from the ACL_GROUP_OBJ entry
  when needed also when the ACL_MASK entry is not re-calculated
  (-n option), so setfacl doesn't complain about an invalid ACL.

0.6.8 (11 October 2000)
* src/walk_tree.c was stat()ing the targets of symbolic links in
  WALK_PHYSICAL mode.
* Removed spurious definitions from src/gettext.h that caused
  problems with several versions of GNU gettext headers
* setfacl calls acl_delete_def_file() if removing the default ACL
  of a directory that doesn't have a default ACL. i(If the user lacks
  appropriate permissions, that no-operation fails.) Fixed.
* Tiny bug in src/showacl script (mask entry not applied to owning
  group entry)

0.6.5 (10 September 2000)
* Command line options revised for POSIX compatibility (now defaults to
  -H half-logical walk, added -L logical walk, -P physical walk).
* Funny post-order option no longer documented.

0.6.0 (24 March 2000)
* Permissions were not set to the union of the ACL_GROUP_OBJ and

libacl/acl_from_text.c
/*
  File: acl_from_text.c

  Copyright (C) 1999, 2000, 2001
  Andreas Gruenbacher, <a.gruenbacher@computer.org>

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU Library General Public
  License as published by the Free Software Foundation; either
  version 2 of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Library General Public License for more details.

  You should have received a copy of the GNU Library General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#include <string.h>
#include <pwd.h>
#include <grp.h>
#include "libacl.h"


#define SKIP_WS(x) do { \
	while (*(x)==' ' || *(x)=='\t' || *(x)=='\n' || *(x)=='\r') \
		(x)++; \
	if (*(x)=='#') { \
		while (*(x)!='\n' && *(x)!='\0') \
			(x)++; \
	} \
	} while (0)


static int parse_acl_entry(const char **text_p, acl_t *acl_p);


/* 23.4.13 */
acl_t
acl_from_text(const char *buf_p)
{
	acl_t acl;
	acl = acl_init(0);
	if (!acl)
		return NULL;
	if (!buf_p) {
		errno = EINVAL;
		return NULL;
	}
	while (*buf_p != '\0') {
		if (parse_acl_entry(&buf_p, &acl) != 0)
			goto fail;
		SKIP_WS(buf_p);
		if (*buf_p == ',') {
			buf_p++;
			SKIP_WS(buf_p);
		}
	}
	if (*buf_p != '\0') {
		errno = EINVAL;
		goto fail;
	}

	return acl;

fail:
	acl_free(acl);
	return NULL;
}


static int
skip_tag_name(const char **text_p, const char *token)
{
	size_t len = strlen(token);
	const char *text = *text_p;

	SKIP_WS(text);
	if (strncmp(text, token, len) == 0) {
		text += len;
		goto delimiter;
	}
	if (*text == *token) {
		text++;
		goto delimiter;
	}
	return 0;

delimiter:
	SKIP_WS(text);
	if (*text == ':')
		text++;
	*text_p = text;
	return 1;
}


static char *
get_token(const char **text_p)
{
	char *token = NULL;
	const char *ep;

	ep = *text_p;
	SKIP_WS(ep);

	while (*ep!='\0' && *ep!='\r' && *ep!='\n' && *ep!=':' && *ep!=',')
		ep++;
	if (ep == *text_p)
		goto after_token;
	token = (char*)malloc(ep - *text_p + 1);
	if (token == 0)
		goto after_token;
	memcpy(token, *text_p, (ep - *text_p));
	token[ep - *text_p] = '\0';
after_token:
	if (*ep == ':')
		ep++;
	*text_p = ep;
	return token;
}


static int
get_id(const char *token, id_t *id_p)
{
	char *ep;
	long l;
	l = strtol(token, &ep, 0);
	if (*ep != '\0')
		return -1;
	if (l < 0) {
		/*
		  Negative values are interpreted as 16-bit numbers,
		  so that id -2 maps to 65534 (nobody/nogroup), etc.
		*/
		l &= 0xFFFF;
	}
	*id_p = l;
	return 0;
}


static int
get_uid(const char *token, uid_t *uid_p)
{
	struct passwd *passwd;

	if (get_id(token, uid_p) == 0)
		goto accept;
	passwd = getpwnam(token);
	if (passwd) {
		*uid_p = passwd->pw_uid;
		goto accept;
	}
	return -1;

accept:
	return 0;
}


static int
get_gid(const char *token, gid_t *gid_p)
{
	struct group *group;

	if (get_id(token, (uid_t *)gid_p) == 0)
		goto accept;
	group = getgrnam(token);
	if (group) {
		*gid_p = group->gr_gid;
		goto accept;
	}
	return -1;

accept:
	return 0;
}


/*
	Parses the next acl entry in text_p.

	Returns:
		-1 on error, 0 on success.
*/

static int
parse_acl_entry(const char **text_p, acl_t *acl_p)
{
	acl_entry_obj entry_obj;
	acl_entry_t entry_d;
	char *str;
	const char *backup;
	int error, perm_chars;

	new_obj_p_here(acl_entry, &entry_obj);
	init_acl_entry_obj(entry_obj);

	/* parse acl entry type */
	SKIP_WS(*text_p);
	switch (**text_p) {
		case 'u':  /* user */
			if (!skip_tag_name(text_p, "user"))
				goto fail;
			backup = *text_p;
			str = get_token(text_p);
			if (str) {
				entry_obj.etag = ACL_USER;
				error = get_uid(str, &entry_obj.eid.qid);
				free(str);
				if (error) {
					*text_p = backup;
					return -1;
				}
			} else {
				entry_obj.etag = ACL_USER_OBJ;
			}
			break;

		case 'g':  /* group */
			if (!skip_tag_name(text_p, "group"))
				goto fail;
			backup = *text_p;
			str = get_token(text_p);
			if (str) {
				entry_obj.etag = ACL_GROUP;
				error = get_gid(str, &entry_obj.eid.qid);
				free(str);
				if (error) {
					*text_p = backup;
					return -1;
				}
			} else {
				entry_obj.etag = ACL_GROUP_OBJ;
			}
			break;

		case 'm':  /* mask */
			if (!skip_tag_name(text_p, "mask"))
				goto fail;
			/* skip empty entry qualifier field (this field may
			   be missing for compatibility with Solaris.) */
			SKIP_WS(*text_p);
			if (**text_p == ':')
				(*text_p)++;
			entry_obj.etag = ACL_MASK;
			break;

		case 'o':  /* other */
			if (!skip_tag_name(text_p, "other"))
				goto fail;
			/* skip empty entry qualifier field (this field may
			   be missing for compatibility with Solaris.) */
			SKIP_WS(*text_p);
			if (**text_p == ':')
				(*text_p)++;
			entry_obj.etag = ACL_OTHER;
			break;

		default:
			goto fail;
	}

	for (perm_chars=0; perm_chars<3; perm_chars++, (*text_p)++) {
		switch(**text_p) {
			case 'r':
				if (entry_obj.eperm.sperm & ACL_READ)
					goto fail;
				entry_obj.eperm.sperm |= ACL_READ;
				break;

			case 'w':
				if (entry_obj.eperm.sperm  & ACL_WRITE)
					goto fail;
				entry_obj.eperm.sperm  |= ACL_WRITE;
				break;

			case 'x':
				if (entry_obj.eperm.sperm  & ACL_EXECUTE)
					goto fail;
				entry_obj.eperm.sperm  |= ACL_EXECUTE;
				break;

			case '-':
				/* ignore */
				break;

			default:
				if (perm_chars == 0)
					goto fail;
				goto create_entry;
		}
	}

create_entry:
	if (acl_create_entry(acl_p, &entry_d) != 0)
		return -1;
	if (acl_copy_entry(entry_d, int2ext(&entry_obj)) != 0)
		return -1;
	return 0;

fail:
	errno = EINVAL;
	return -1;
}

